File: dotnet-server/Services/WebhookService.cs
using System.Text.Json;
using Microsoft.Extensions.Options;
using PaymentGateway.Api.Config;
using PaymentGateway.Api.Repositories;
using PaymentGateway.Api.Utilities;

namespace PaymentGateway.Api.Services;

public interface IWebhookService
{
    Task ProcessWebhookAsync(string rawBody, string signature);
}

public class WebhookService : IWebhookService
{
    private readonly RazorpaySettings _settings;
    private readonly IPaymentRepository _repository;
    private readonly ILogger<WebhookService> _logger;

    public WebhookService(
        IOptions<RazorpaySettings> settings,
        IPaymentRepository repository,
        ILogger<WebhookService> logger)
    {
        _settings = settings.Value;
        _repository = repository;
        _logger = logger;
    }

    public async Task ProcessWebhookAsync(string rawBody, string signature)
    {
        if (string.IsNullOrEmpty(signature))
        {
            throw new BadRequestException("Missing x-razorpay-signature header.");
        }

        var isValid = CryptoUtils.VerifyWebhookSignature(rawBody, signature, _settings.WebhookSecret);
        if (!isValid)
        {
            _logger.LogError("Webhook HMAC signature verification failed.");
            throw new PaymentVerificationException("Invalid webhook signature.");
        }

        using var doc = JsonDocument.Parse(rawBody);
        var root = doc.RootElement;

        var eventType = root.GetProperty("event").GetString() ?? "";
        var eventId = root.TryGetProperty("event_id", out var evtId) ? evtId.GetString() ?? Guid.NewGuid().ToString() : Guid.NewGuid().ToString();

        _logger.LogInformation("Received verified Webhook event: {EventType} [ID: {EventId}]", eventType, eventId);

        if (eventType is "payment.captured" or "order.paid")
        {
            if (root.TryGetProperty("payload", out var payload) && payload.TryGetProperty("payment", out var paymentObj))
            {
                var entity = paymentObj.GetProperty("entity");
                var orderId = entity.GetProperty("order_id").GetString() ?? "";
                var paymentId = entity.GetProperty("id").GetString() ?? "";

                await _repository.RecordWebhookEventAsync(orderId, eventId, eventType);
                await _repository.UpdateStatusAsync(orderId, "completed", paymentId);
            }
        }
        else if (eventType is "payment.failed")
        {
            if (root.TryGetProperty("payload", out var payload) && payload.TryGetProperty("payment", out var paymentObj))
            {
                var entity = paymentObj.GetProperty("entity");
                var orderId = entity.GetProperty("order_id").GetString() ?? "";
                var reason = entity.TryGetProperty("error_description", out var desc) ? desc.GetString() : "Payment failed";

                await _repository.RecordWebhookEventAsync(orderId, eventId, eventType);
                await _repository.UpdateStatusAsync(orderId, "failed", failureReason: reason);
            }
        }
    }
}

File: server/src/services/webhookService.js
import { envConfig } from "../config/env.js";
import { CryptoUtils } from "../utils/cryptoUtils.js";
import { paymentRepository } from "../repositories/paymentRepository.js";
import { BadRequestError, PaymentVerificationError } from "../utils/AppError.js";
import { logger } from "../utils/logger.js";

/**
 * Webhook Service: Asynchronous Razorpay Event Processor.
 * Verifies webhook signatures and handles events (`payment.captured`, `payment.failed`, `order.paid`).
 */
export class WebhookService {
  /**
   * Process incoming Razorpay webhook payload safely and idempotently
   * @param {Buffer|string} rawBody - Raw unparsed HTTP payload
   * @param {string} signature - Header x-razorpay-signature
   * @param {Object} eventPayload - Parsed webhook event JSON
   */
  async processWebhookEvent(rawBody, signature, eventPayload) {
    if (!signature) {
      throw new BadRequestError("Missing x-razorpay-signature header in webhook request.");
    }

    const webhookSecret = envConfig.RAZORPAY_WEBHOOK_SECRET;
    const isValid = CryptoUtils.verifyWebhookSignature(rawBody, signature, webhookSecret);

    if (!isValid) {
      logger.error("Webhook signature verification failed.");
      throw new PaymentVerificationError("Invalid webhook signature.");
    }

    const { event, payload } = eventPayload;
    const eventId = eventPayload.event_id || `evt_${Date.now()}`;

    logger.info(`Received verified Webhook event: ${event} [ID: ${eventId}]`);

    switch (event) {
      case "payment.captured":
      case "order.paid":
        await this.handlePaymentSuccess(payload, eventId, event);
        break;

      case "payment.failed":
        await this.handlePaymentFailure(payload, eventId, event);
        break;

      default:
        logger.info(`Unhandled Webhook event type: ${event}`);
        break;
    }

    return { processed: true, event };
  }

  async handlePaymentSuccess(payload, eventId, eventType) {
    const entity = payload.payment ? payload.payment.entity : payload.order.entity;
    const orderId = entity.order_id || entity.id;
    const paymentId = entity.id;

    logger.info(`Webhook Processing: Payment Success for Order ${orderId}`);

    // Idempotent recording
    await paymentRepository.recordWebhookEvent(orderId, eventId, eventType);
    await paymentRepository.updateStatus(orderId, {
      razorpayPaymentId: paymentId,
      status: "completed",
    });
  }

  async handlePaymentFailure(payload, eventId, eventType) {
    const entity = payload.payment.entity;
    const orderId = entity.order_id;

    logger.warn(`Webhook Processing: Payment Failed for Order ${orderId}`);

    await paymentRepository.recordWebhookEvent(orderId, eventId, eventType);
    await paymentRepository.updateStatus(orderId, {
      status: "failed",
      failureReason: entity.error_description || "Payment failed at gateway",
    });
  }
}

export const webhookService = new WebhookService();
